File: app/src/main/java/workshop1024/com/xproject/controller/activity/MainActivity.java
package workshop1024.com.xproject.controller.activity;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.NavigationView;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.Toast;

import workshop1024.com.xproject.R;
import workshop1024.com.xproject.controller.fragment.TopFragment;
import workshop1024.com.xproject.controller.fragment.XFragment;
import workshop1024.com.xproject.controller.fragment.home.NewsListFragment;
import workshop1024.com.xproject.controller.fragment.home.HomePageFragment;
import workshop1024.com.xproject.controller.fragment.save.SavedFragment;
import workshop1024.com.xproject.view.BottomMenu;

/**
 * 主页面，包含抽屉导航栏，以及导航菜单对应的各个子Fragment页面
 */
public class MainActivity extends AppCompatActivity implements View.OnClickListener, SearchView.OnQueryTextListener,
        NavigationView.OnNavigationItemSelectedListener, FragmentManager.OnBackStackChangedListener {
    //抽屉导航区域
    //抽屉视图
    private DrawerLayout mDrawerLayut;
    //抽屉列表视图
    private NavigationView mNavigationView;

    //登陆按钮
    private Button mLoginButton;
    //退出按钮
    private ImageButton mLogoutButton;

    //内容区域
    private CoordinatorLayout mRightCoordinatorLayout;
    //工具栏
    private Toolbar mToolbar;
    private ActionBarDrawerToggle mActionBarDrawerToggle;

    private FragmentManager mFragmentManager;
    private XFragment mCurrentFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_activity);

        mDrawerLayut = findViewById(R.id.main_drawerlayout_navigator);
        mNavigationView = findViewById(R.id.mainleft_navigationview);
        mRightCoordinatorLayout = findViewById(R.id.mainright_coordinatorlayout_root);
        mToolbar = findViewById(R.id.main_toolbar_navigator);


        View headerLayout = mNavigationView.inflateHeaderView(R.layout.mainleft_navigator_header);
        mLoginButton = headerLayout.findViewById(R.id.leftnavigator_button_login);
        mLogoutButton = headerLayout.findViewById(R.id.leftnavigator_imagebutton_logout);

        setSupportActionBar(mToolbar);
        registerForContextMenu(findViewById(R.id.mainright_floatingactionbutton_action));
        //创建ActionBar左边的up action，点击开关左侧抽屉导航
        mActionBarDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayut, mToolbar, R
                .string.navigation_drawer_open, R.string.navigation_drawer_close);
        mDrawerLayut.addDrawerListener(mActionBarDrawerToggle);
        mActionBarDrawerToggle.syncState();
        mActionBarDrawerToggle.setToolbarNavigationClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });
        mActionBarDrawerToggle.setDrawerSlideAnimationEnabled(false);
        mNavigationView.setNavigationItemSelectedListener(this);

        mLoginButton.setOnClickListener(this);
        mLogoutButton.setOnClickListener(this);

        mFragmentManager = getSupportFragmentManager();
        mFragmentManager.addOnBackStackChangedListener(this);

        showHomePageFragment();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();

        //根据当前显示的Fragment的类型，更新显示的菜单
        if (mCurrentFragment instanceof HomePageFragment) {
            menuInflater.inflate(R.menu.homepage_toolbar_menu, menu);
            setTitle(R.string.toolbar_title_home);
            SearchView searchView = (SearchView) menu.findItem(R.id.main_menu_search).getActionView();
            searchView.setOnQueryTextListener(this);
        } else if (mCurrentFragment instanceof NewsListFragment) {
            menuInflater.inflate(R.menu.homelist_toolbar_menu, menu);
        } else if (mCurrentFragment instanceof SavedFragment) {
            menuInflater.inflate(R.menu.saved_toolbar_menu, menu);
            setTitle(R.string.toolbar_title_saved);
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.main_menu_search:
                Toast.makeText(this, "toolbar_menu_search", Toast.LENGTH_SHORT).show();
                break;
            case R.id.main_menu_add:
                BottomMenu bottomMenu = new BottomMenu(this);
                bottomMenu.showAtLocation(mRightCoordinatorLayout, Gravity.BOTTOM, 0, 0);
                break;
            case R.id.main_menu_refresh:
                Toast.makeText(this, "main_menu_refresh", Toast.LENGTH_SHORT).show();
//            mHomeFragment.onRefresh();
                break;
            case R.id.main_menu_marked:
                Toast.makeText(this, "main_menu_marked", Toast.LENGTH_SHORT).show();
                break;
            case R.id.main_menu_rate:
                Toast.makeText(this, "main_menu_rate", Toast.LENGTH_SHORT).show();
                break;
            case R.id.main_menu_product:
                Toast.makeText(this, "main_menu_product", Toast.LENGTH_SHORT).show();
                break;
            case R.id.main_menu_about:
                Toast.makeText(this, "main_menu_about", Toast.LENGTH_SHORT).show();
                break;
            case R.id.main_menu_cards:
                Toast.makeText(this, "main_menu_cards", Toast.LENGTH_SHORT).show();
                //FIXME 实现有点别扭
                ((NewsListFragment) mCurrentFragment).showBigCardsList();
                break;
            case R.id.main_menu_compact:
                Toast.makeText(this, "main_menu_compact", Toast.LENGTH_SHORT).show();
                ((NewsListFragment) mCurrentFragment).showCompactList();
                break;
            case R.id.main_menu_minimal:
                Toast.makeText(this, "main_menu_minimal", Toast.LENGTH_SHORT).show();
                ((NewsListFragment) mCurrentFragment).showMinimalList();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        int selectItemId = item.getItemId();

        if (selectItemId == R.id.leftnavigator_menu_home) {
            showHomePageFragment();
        } else if (selectItemId == R.id.leftnavigator_menu_saved) {
            SavedFragment savedFragment = SavedFragment.newInstance();
            mFragmentManager.beginTransaction().replace(R.id.mainright_framelayout_fragments, savedFragment).commit();

            //没有添加到Fragment堆栈管理，则需要单独处理当前显示的Fragment，导航列表选项逻辑
            mNavigationView.setCheckedItem(R.id.leftnavigator_menu_saved);
            mCurrentFragment = savedFragment;
        } else if (selectItemId == R.id.leftnavigator_menu_settings) {
            Intent intent = new Intent(this, SettingsActivity.class);
            startActivity(intent);
        } else if (selectItemId == R.id.leftnavigator_menu_feedback) {
            Intent intent = new Intent(this, FeedbackActivity.class);
            startActivity(intent);
        }

        mDrawerLayut.closeDrawer(GravityCompat.START);
        invalidateOptionsMenu();

        return true;
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        mActionBarDrawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mActionBarDrawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    public void onBackStackChanged() {
        mCurrentFragment = (XFragment) mFragmentManager.findFragmentById(R.id.mainright_framelayout_fragments);
        //当MainActivity还显示Fragment的之后执行相关逻辑，MainActivity不展示Fragment相关逻辑
        if (mCurrentFragment != null) {
            //更具当前显示的Fragment多包含的导航栏id，更新导航栏列表选中的选项
            mNavigationView.setCheckedItem(mCurrentFragment.getNavigationItemId());

            //如果是一级Fragment则显示抽屉导航图标，如果只其它级别Fragment这显示返回上一页图标
            mActionBarDrawerToggle.setDrawerIndicatorEnabled(mCurrentFragment instanceof TopFragment ? true : false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(mCurrentFragment instanceof TopFragment ? false : true);
            mActionBarDrawerToggle.syncState();

            //Fragment堆栈有变化，根据当前显示的Fragment重新更新菜单展示
            invalidateOptionsMenu();
        }
    }

    @Override
    public void onClick(View view) {
        if (view == mLoginButton) {
            Intent intent = new Intent(MainActivity.this, LoginActivity.class);
            startActivity(intent);
        } else if (view == mLogoutButton) {

        }
    }

    private void showHomePageFragment() {
        HomePageFragment homePageFragment = HomePageFragment.newInstance();
        mFragmentManager.beginTransaction().replace(R.id.mainright_framelayout_fragments, homePageFragment).commit();

        //没有添加到Fragment堆栈管理，则需要单独处理当前显示的Fragment，导航列表选项逻辑
        mNavigationView.setCheckedItem(R.id.leftnavigator_menu_home);
        mCurrentFragment = homePageFragment;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        NewsListFragment newsListFragment = NewsListFragment.newInstance("Search", query);
        mFragmentManager.beginTransaction().replace(R.id.mainright_framelayout_fragments, newsListFragment)
                .addToBackStack("").commit();
        setTitle(query);
        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        Log.i("XProject", "onQueryTextChange =" + newText);
        return false;
    }
}

File: main/src/main/java/workshop1024/com/xproject/main/controller/activity/MainActivity.kt
package workshop1024.com.xproject.main.controller.activity

import android.content.ComponentName
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import androidx.core.view.GravityCompat
import android.util.Log
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.FragmentManager
import com.alibaba.android.arouter.facade.annotation.Route
import com.alibaba.android.arouter.launcher.ARouter
import com.google.android.material.navigation.NavigationView
import org.greenrobot.eventbus.EventBus
import workshop1024.com.xproject.base.arouter.provider.SaveProvider
import workshop1024.com.xproject.base.controller.event.*
import workshop1024.com.xproject.base.controller.fragment.TopFragment
import workshop1024.com.xproject.base.controller.fragment.XFragment
import workshop1024.com.xproject.base.exception.IntentNotFoundException
import workshop1024.com.xproject.base.service.ServiceFactory
import workshop1024.com.xproject.base.utils.IntentUtils
import workshop1024.com.xproject.base.utils.ReflectUtils
import workshop1024.com.xproject.main.R
import workshop1024.com.xproject.main.databinding.MainActivityBinding
import workshop1024.com.xproject.main.databinding.MainLeftnavigatorHeaderBinding
import workshop1024.com.xproject.main.view.popupwindow.BottomMenu
import kotlin.jvm.internal.Reflection

/**
 * 主页面，包含抽屉导航栏，以及导航菜单对应的各个子Fragment页面
 */
//ARouter：在支持路由的页面上添加注解(必选) ，这里的路径需要注意的是至少需要有两级，/xx/xx
@Route(path = "/main/MainActivity")
class MainActivity : AppCompatActivity(), SearchView.OnQueryTextListener, FragmentManager.OnBackStackChangedListener {
    private lateinit var mActionBarDrawerToggle: ActionBarDrawerToggle

    private var mCurrentFragment: XFragment? = null

    private lateinit var mMainActivityBinding: MainActivityBinding


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mMainActivityBinding = DataBindingUtil.setContentView(this, R.layout.main_activity)
        mMainActivityBinding.mainHandlers = MainHandlers()

        val mainLeftNavigatorHeaderBinding = DataBindingUtil.inflate<MainLeftnavigatorHeaderBinding>(layoutInflater,
                R.layout.main_leftnavigator_header, mMainActivityBinding.mainleftNavigationview, false)
        mainLeftNavigatorHeaderBinding.headerHandlers = HeaderHandlers()

        //设置抽屉导航HeaderView视图
        mMainActivityBinding.mainleftNavigationview.addHeaderView(mainLeftNavigatorHeaderBinding.root)

        //设置顶部toolBar视图
        setSupportActionBar(mMainActivityBinding.mainIncludeRight.mainrightToolbarNavigator)

        //创建ActionBar左边的up action，点击开关左侧抽屉导航
        mActionBarDrawerToggle = ActionBarDrawerToggle(this, mMainActivityBinding.mainDrawerlayoutNavigator,
                mMainActivityBinding.mainIncludeRight.mainrightToolbarNavigator, R.string.navigation_drawer_open,
                R.string.navigation_drawer_close)
        mMainActivityBinding.mainDrawerlayoutNavigator.addDrawerListener(mActionBarDrawerToggle)
        mActionBarDrawerToggle.syncState()
        mActionBarDrawerToggle.setToolbarNavigationClickListener {
            View.OnClickListener {
                onBackPressed()
            }
        }
        mActionBarDrawerToggle.isDrawerSlideAnimationEnabled = false

        //处理Fragment的交互显示，默认显示HomeFragment
        supportFragmentManager.addOnBackStackChangedListener(this)
        showHomePageFragment()
    }

    private fun showHomePageFragment() {
        //反射动态获取homePageFragment实例对象，避免对HomePageFragment类的直接依赖
        try {
            //Caused by: f.a.a.b.i.b: 你调用的方法:newInstance，所在的类:workshop1024.com.xproject.home.controller.fragment.HomePageFragment不存在!
            //反射调用，需要配置混淆忽

            //FIXME 已经配置了混淆忽略，发射还继续报错!!
            //Caused by: a.d
            //        at workshop1024.com.xproject.base.utils.ReflectUtils.java.lang.Object invokeCompanionMethod(java.lang.String,java.lang.String,java.lang.Object[])(:7)
            val homePageFragment = ReflectUtils.invokeCompanionMethod("workshop1024.com.xproject.home.controller.fragment.HomePageFragment",
                    "newInstance", R.id.leftnavigator_menu_home) as XFragment

            supportFragmentManager.beginTransaction().replace(R.id.mainright_framelayout_fragments, homePageFragment).commit()

            //没有添加到Fragment堆栈管理，则需要单独处理当前显示的Fragment，导航列表选项逻辑
            mMainActivityBinding.mainleftNavigationview.setCheckedItem(R.id.leftnavigator_menu_home)
            mCurrentFragment = homePageFragment
        } catch (exception: ClassNotFoundException) {
            Log.e("XProject", "反射获取homePageFragment实例失败，调用类路径不存在！")
            exception.toString()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        //根据当前显示的Fragment的类型，更新显示的菜单
        //通过公共组件接口方式，实现组件之间的通信
        //FIXME ? !!等如何合理的设置和配置，不能总是靠提示让代码勉强运行
        if (ServiceFactory.getInstance()?.homeService?.isHomePageFragment(mCurrentFragment)!!) {
            menuInflater.inflate(R.menu.homepage_toolbar_menu, menu)
            setTitle(R.string.toolbar_title_home)
            val searchView = menu?.findItem(R.id.homepage_menu_search)?.actionView as SearchView
            searchView.setOnQueryTextListener(this)
        } else if (ServiceFactory.getInstance()?.homeService?.isNewsListFragment(mCurrentFragment)!!) {
            menuInflater.inflate(R.menu.newslist_toolbar_menu, menu)
        }

        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.homepage_menu_add -> {
                val bottomMenu = BottomMenu(this)
                bottomMenu.showAtLocation(mMainActivityBinding.mainIncludeRight.mainrightCoordinatorlayoutRoot, Gravity.BOTTOM, 0, 0)
            }
            R.id.homepage_menu_refresh -> EventBus.getDefault().post(HomePageRefreshEvent())
            R.id.homepage_menu_marked -> EventBus.getDefault().post(HomePageAsReadEvent())
            R.id.homepage_menu_product -> {
                try {
                    IntentUtils.startActivityByNewIntent(this, "workshop1024.com.xproject.introduce.controller.activity.IntroduceActivity")
                } catch (exception: IntentNotFoundException) {
                    exception.printStackTrace();
                    Log.i("XProject", "跳转介绍页面的意图不存在！")
                }
            }
            R.id.homepage_menu_about -> AboutActivity.startActivity(this)
            R.id.newslist_menu_cards -> EventBus.getDefault().post(NewsListShowBigCardsEvent())
            R.id.newslist_menu_compact -> EventBus.getDefault().post(NewsListShowCompactEvent())
            R.id.newslist_menu_minimal -> EventBus.getDefault().post(NewsListShowMinimalEvent())
            R.id.newslist_menu_marked -> EventBus.getDefault().post(NewsListAsReadEvent())
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        mActionBarDrawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        mActionBarDrawerToggle.onConfigurationChanged(newConfig)
    }

    override fun onBackStackChanged() {
        mCurrentFragment = supportFragmentManager.findFragmentById(R.id.mainright_framelayout_fragments) as XFragment?
        //当MainActivity还显示Fragment的之后执行相关逻辑，MainActivity不展示Fragment相关逻辑

        if (mCurrentFragment != null) {
            //更具当前显示的Fragment多包含的导航栏id，更新导航栏列表选中的选项
            mMainActivityBinding.mainleftNavigationview.setCheckedItem(mCurrentFragment!!.mNavigationItemId)

            //如果是一级Fragment则显示抽屉导航图标和隐藏FloatingActionButton，如果只其它级别Fragment这显示返回上一页图标
            mActionBarDrawerToggle.isDrawerSlideAnimationEnabled = mCurrentFragment is TopFragment
            supportActionBar?.setDisplayHomeAsUpEnabled(mCurrentFragment !is TopFragment)
            mActionBarDrawerToggle.syncState()

            mMainActivityBinding.mainIncludeRight.mainrightFloatingactionbuttonAction.visibility = if (mCurrentFragment is TopFragment) View.GONE
            else View.VISIBLE

            //Fragment堆栈有变化，根据当前显示的Fragment重新更新菜单展示
            invalidateOptionsMenu()
        }
    }


    override fun onQueryTextSubmit(query: String?): Boolean {
        //FIXME 智能通过类型转换吗？？
        val searchNewsFragment = query?.let {
            ReflectUtils.invokeCompanionMethod("workshop1024.com.xproject.home.controller.fragment.news.SearchNewsFragment",
                    "newInstance", it, R.id.leftnavigator_menu_home)
        } as XFragment
        supportFragmentManager.beginTransaction().replace(R.id.mainright_framelayout_fragments, searchNewsFragment).addToBackStack("").commit()
        title = query

        return false
    }

    override fun onQueryTextChange(newText: String?): Boolean {
        return false
    }


    inner class MainHandlers : NavigationView.OnNavigationItemSelectedListener {
        fun onClickAction(view: View) {
            EventBus.getDefault().post(NewsListAsReadEvent())
        }

        override fun onNavigationItemSelected(item: MenuItem): Boolean {
            when (item.itemId) {
                R.id.leftnavigator_menu_home -> showHomePageFragment()
                R.id.leftnavigator_menu_saved -> {
                    //动态创建Fragment:ARouter，跨组件创建savedFragment实例
                    //使用依赖查找发现线服务，通过SaveProvider::class.java类型查找到saveProvider如无
                    val saveProvider = ARouter.getInstance().navigation(SaveProvider::class.java)
                    val savedFragment = saveProvider.newSavedFragmentInstance(R.id.leftnavigator_menu_saved)
                    supportFragmentManager.beginTransaction().replace(R.id.mainright_framelayout_fragments, savedFragment).commit()

                    //没有添加到Fragment堆栈管理，则需要单独处理当前显示的Fragment，导航列表选项逻辑
                    mMainActivityBinding.mainleftNavigationview.setCheckedItem(R.id.leftnavigator_menu_saved)
                    mCurrentFragment = savedFragment as XFragment
                }
                R.id.leftnavigator_menu_settings -> {
                    //隐式意图2-setClass(content,...)，通过公共组件实现依赖
                    try {
                        IntentUtils.startActivityBySetClassName(this@MainActivity, "workshop1024.com.xproject.settings.controller.activity.SettingsActivity")
                    } catch (exception: IntentNotFoundException) {
                        exception.printStackTrace()
                        Log.e("XProject", "跳转SettingsActivity的意图不存在！")
                    }
                }
                R.id.leftnavigator_menu_feedback -> {
                    //隐式意图3-setComponent，避免组件直接依赖
                    try {
                        IntentUtils.startActivityBySetComponent(this@MainActivity, "workshop1024.com.xproject.feedback.controller.activity.FeedbackActivity")
                    } catch (exception: IntentNotFoundException) {
                        exception.printStackTrace()
                        Log.e("XProject", "跳转FeedbackActivity的意图不存在！")
                    }
                }
            }

            mMainActivityBinding.mainDrawerlayoutNavigator.closeDrawer(GravityCompat.START)
            invalidateOptionsMenu()

            return true
        }

    }

    inner class HeaderHandlers {
        fun onClickLogin(view: View) {
            //ARouter：应用内简单的跳转，build填写地址，navigation发射路由
            ARouter.getInstance().build("/login/LoginActivity").navigation()
        }

        fun onClickLogout(view: View) {
            Log.i("XProject", "onClickLogout")
        }
    }
}